finetune/encoding/target_encoders.py

```diff
@@ -239,21 +300,461 @@ def transform(self, out, labels):
+        
+    def transform(self, out, labels):
+        labels, groups = labels
+
+        # Add a group_start value to every label for the sequence labeler
+        # None -> Not in group, True -> Start of group, False -> In group
+        # This could potentially be pulled out of the sequence labeler
+        for label in labels:
+            label["group_start"] = None
+        for group in groups:
+            if not is_continuous(group):
+                continue
+
+            # Gather entities inside group
+            # Note that we assume groups are made up of only entities
+            group_start = min([t["start"] for t in group["tokens"]])
+            group_end = max([t["end"] for t in group["tokens"]])
+            group_labels = []
+            for label in labels:
+                if label["start"] >= group_start and label["end"] <= group_end:
+                    group_labels.append(label)
+            # Add group_start information
+            if group_labels:
+                group_labels = sorted(group_labels, key=lambda x: x["start"])
+                group_labels[0]["group_start"] = True
+                for label in group_labels[1:]:
+                    label["group_start"] = False
+
+        ret = super().transform(out, labels)
+        # Un-do the changes to the underlying data
+        for label in labels:
+            del label["group_start"]
+        return ret
+
+class MultiCRFGroupSequenceLabelingEncoder(SequenceLabelingEncoder):
+    """
+    Produces labels for a model with a seperate CRF for NER and group
+    information.
+
+    Outputs two label sequences, the first for the standard NER CRF and the
+    second for the group CRF.
+    """
+    def __init__(self, pad_token, bio_tagging=True):
+        super().__init__(pad_token, bio_tagging=bio_tagging, group_tagging=True)
+
+    def fit(self, labels):
+        labels, groups = list(zip(*labels))
+        super().fit(labels)
+
+        self.label_classes_ = self.classes_
+        self.label_lookup = self.lookup
+
+        self.group_classes_ = [self.pad_token, "BG-", "IG-"]
+        self.group_lookup = {c: i for i, c in enumerate(self.group_classes_)}
+        
+    def transform(self, out, labels):
+        labels, groups = labels
+
+        # Create "fake" labels for groups
+        group_labels = []
+        for group in groups:
+            if not is_continuous(group):
+                continue
+            group_start = min([t["start"] for t in group["tokens"]])
+            group_end = max([t["end"] for t in group["tokens"]])
+            group_text = " ".join([t["text"] for t in group["tokens"]])
+            group_labels.append({
+                "start": group_start,
+                "end": group_end,
+                "label": "",
+                "text": group_text,
+                "group_start": True,
+            })
+
+        # Encode grouping labels
+        self.classes_ = self.group_classes_
```

Comment:
The set state on self --> call super() pattern is a little bit gross, part of me thinks this would be cleaner if it were more functional  + settings like bio_tagging / group_tagging could be passed to .transform().

Reply:
Totally agree that its gross, but was mainly trying to avoid adding more code to the Sequence Labeling Encoder. Would have liked to come up with a better solution though